File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/AddressComponentsString.java
package spectable.common;

import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class AddressComponentsString {
    private static final String DNCString = "?DNC?";

    public String zip;
    public String streetName;
    public String city;
    public String preDirection;
    public String suffixDirection;
    public String state;
    public String suffixType;

    public AddressComponentsString(String zip, String streetName, String city, String preDirection, String suffixDirection, String state, String suffixType) {
        this.zip = zip;
        this.streetName = streetName;
        this.city = city;
        this.preDirection = preDirection;
        this.suffixDirection = suffixDirection;
        this.state = state;
        this.suffixType = suffixType;
    }

    /** Builds from the text form, e.g. Money as "25 USD". */
    public static AddressComponentsString fromText(String text) {
        java.util.List<String> parts = Tokens.require(text, 7, "AddressComponents");
        return new AddressComponentsString(parts.get(0), parts.get(1), parts.get(2), parts.get(3), parts.get(4), parts.get(5), parts.get(6));
    }

    public AddressComponentsString(String text) {
        AddressComponentsString parsed = fromText(text);
        this.zip = parsed.zip;
        this.streetName = parsed.streetName;
        this.city = parsed.city;
        this.preDirection = parsed.preDirection;
        this.suffixDirection = parsed.suffixDirection;
        this.state = parsed.state;
        this.suffixType = parsed.suffixType;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof AddressComponentsString)) return false;
        AddressComponentsString that = (AddressComponentsString) o;
        return (DNCString.equals(zip) || DNCString.equals(that.zip) || Objects.equals(zip, that.zip))
            && (DNCString.equals(streetName) || DNCString.equals(that.streetName) || Objects.equals(streetName, that.streetName))
            && (DNCString.equals(city) || DNCString.equals(that.city) || Objects.equals(city, that.city))
            && (DNCString.equals(preDirection) || DNCString.equals(that.preDirection) || Objects.equals(preDirection, that.preDirection))
            && (DNCString.equals(suffixDirection) || DNCString.equals(that.suffixDirection) || Objects.equals(suffixDirection, that.suffixDirection))
            && (DNCString.equals(state) || DNCString.equals(that.state) || Objects.equals(state, that.state))
            && (DNCString.equals(suffixType) || DNCString.equals(that.suffixType) || Objects.equals(suffixType, that.suffixType));
    }

    @Override
    public int hashCode() {
        return Objects.hash(zip, streetName, city, preDirection, suffixDirection, state, suffixType);
    }

    @Override
    public String toString() {
        return Tokens.token(zip) + " " + Tokens.token(streetName) + " " + Tokens.token(city) + " " + Tokens.token(preDirection) + " " + Tokens.token(suffixDirection) + " " + Tokens.token(state) + " " + Tokens.token(suffixType);
    }
}

File: JavaScript/ExampleTests/common/AddressComponentsString.js
import * as tokens from "./tokens.js";

export class AddressComponentsString {
  static DNC_STRING = "?DNC?";

  constructor(zip = "", streetName = "", city = "", preDirection = "", suffixDirection = "", state = "", suffixType = "") {
    this.zip = zip;
    this.streetName = streetName;
    this.city = city;
    this.preDirection = preDirection;
    this.suffixDirection = suffixDirection;
    this.state = state;
    this.suffixType = suffixType;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new AddressComponentsString(
      v[0] ?? "",
      v[1] ?? "",
      v[2] ?? "",
      v[3] ?? "",
      v[4] ?? "",
      v[5] ?? "",
      v[6] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 7, "AddressComponents");
    return new AddressComponentsString(
      parts[0],
      parts[1],
      parts[2],
      parts[3],
      parts[4],
      parts[5],
      parts[6]
    );
  }

  toString() {
    return tokens.token(this.zip) + " " + tokens.token(this.streetName) + " " + tokens.token(this.city) + " " + tokens.token(this.preDirection) + " " + tokens.token(this.suffixDirection) + " " + tokens.token(this.state) + " " + tokens.token(this.suffixType);
  }

  equals(other) {
    if (!(other instanceof AddressComponentsString)) return false;
    return (this.zip === AddressComponentsString.DNC_STRING || other.zip === AddressComponentsString.DNC_STRING || this.zip === other.zip)
      && (this.streetName === AddressComponentsString.DNC_STRING || other.streetName === AddressComponentsString.DNC_STRING || this.streetName === other.streetName)
      && (this.city === AddressComponentsString.DNC_STRING || other.city === AddressComponentsString.DNC_STRING || this.city === other.city)
      && (this.preDirection === AddressComponentsString.DNC_STRING || other.preDirection === AddressComponentsString.DNC_STRING || this.preDirection === other.preDirection)
      && (this.suffixDirection === AddressComponentsString.DNC_STRING || other.suffixDirection === AddressComponentsString.DNC_STRING || this.suffixDirection === other.suffixDirection)
      && (this.state === AddressComponentsString.DNC_STRING || other.state === AddressComponentsString.DNC_STRING || this.state === other.state)
      && (this.suffixType === AddressComponentsString.DNC_STRING || other.suffixType === AddressComponentsString.DNC_STRING || this.suffixType === other.suffixType);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}
